### ISSUE TITLE
TorchScripted BERT model not working on GPU

### ISSUE THREAD

**Issue by chauhang** [pytorch/serve#479]
The BERT model [example](https://github.com/pytorch/serve/tree/master/examples/Huggingface_Transformers) is not working on GPU machines. Model gets loaded but the predictions are failing.

## Context
* torchserve version: Tested with both v0.1.1 release and install from source 
* torch version: 1.5.1
* torchvision version [if any]: 0.6.1
* torchtext version [if any]: 0.6.0
* torchaudio version [if any]: n/a
* java version: 11
* Operating System and version: ubuntu 18.04
* Transformers: v2.11.0

## Your Environment
* Installed using source? [yes/no]: tested with both - same issue on both
* Are you planning to deploy it using docker container? [yes/no]: no
* Is it a CPU or GPU environment?: GPU
* Using a default/custom handler? [If possible upload/share custom handler/model]: BERT model example
* What kind of model is it e.g. vision, text, audio?: Transformer model
* Are you planning to use local models from model-store or public url being used e.g. from S3 bucket etc.?
  [If public url then provide link.]: S3
* Provide config.properties, logs [ts.log] and parameters used for model registration/update APIs:
* Link to your project [if any]:

## Expected Behavior
Prediction example using steps in readme should work `http://127.0.0.1:8080/predictions/my_tc -T ./Seq_classification_artifacts/sample_text.txt`

## Current Behavior
Getting 503 error instead. See detailed logs: 

## Possible Solution
None for GPU at present

## Steps to Reproduce
Followed steps in readme
1. python Download_Transformer_models.py
2. torch-model-archiver --model-name BERTSeqClassification_Torchscript --version 1.0 --serialized-file Transformer_model/traced_model.pt --handler ./Transformer_handler_generalized.py --extra-files "./setup_config.json,./Seq_classification_artifacts/index_to_name.json"
3. mkdir model_store
4. mv BERTSeqClassification_Torchscript.mar model_store/
5. torchserve --start --model-store model_store --models my_tc=BERTSeqClassification_Torchscript.mar
6. http://127.0.0.1:8080/predictions/my_tc -T ./Seq_classification_artifacts/sample_text.txt

## Failure Logs [if any]
Error log attached
[bert-error.txt](https://github.com/pytorch/serve/files/4827197/bert-error.txt)

**Comment by vdantu** [comment]
Was the model scripted on a CPU? If yes, it could be related to [this](https://github.com/huggingface/transformers/pull/5773/files)

**Comment by maaquib** [comment]
Pending merge of https://github.com/huggingface/transformers/pull/5773

**Comment by mfuntowicz** [comment]
@maaquib The PR has been upstream on transformers 👍.

Let us know may you have any follow up issue(s)/question(s)

Morgan from 🤗

**Comment by HamidShojanazeri** [comment]
Hi @mfuntowicz, thank you for pushing the [PR](https://github.com/huggingface/transformers/pull/5773) and working on this issue. I am trying to trace/save the bert_base model for sequence classification on GPU and then load and run it on CPU, and I am facing the issue which is not about position_embedding anymore. I tried the following code (device is set to cuda) and it complains about token_type_embeddings, however, if I pass the attention_mask and token_type_ids,"(traced_model = torch.jit.trace(model, (input_ids,attention_mask,token_type_ids)))" when tracing, it would pass. I would appreciate your thoughts on this, thanks. 

```
import transformers
import os
import json
import torch
from transformers import (AutoModelForSequenceClassification, AutoTokenizer, AutoModelForQuestionAnswering,
 AutoModelForTokenClassification, AutoConfig)
print('Transformers version',transformers.__version__)
device = torch.device('cuda' if torch.cuda.is_available() else 'cpu')
dummy_input = "This is a dummy input for torch jit trace"
max_length=20
config = AutoConfig.from_pretrained('bert-base-uncased',num_labels=2,torchscript=True)
model = AutoModelForSequenceClassification.from_pretrained('bert-base-uncased', config=config)
tokenizer = AutoTokenizer.from_pretrained('bert-base-uncased',do_lower_case=True)
inputs = tokenizer.encode_plus(dummy_input,max_length = max_length,pad_to_max_length = True,truncation=True, add_special_tokens = True, return_tensors = 'pt')

print(inputs.keys())
input_ids = inputs["input_ids"].to(device)
attention_mask = inputs["attention_mask"].to(device)
token_type_ids = inputs["token_type_ids"].to(device)
model.to(device).eval()

traced_model = torch.jit.trace(model, (input_ids,attention_mask,token_type_ids))
torch.jit.save(traced_model, "BERT_cuda.pt")

print(traced_model.graph)
print("\n")
print("Load model onto CPU")
loaded = torch.jit.load("BERT_cuda.pt", map_location=torch.device("cpu"))
print("\n")
print(loaded.graph)
outputs = loaded(input_ids.to("cpu"),attention_mask.to("cpu"),token_type_ids.to("cpu"))
print(outputs)
```
##Error

Traceback (most recent call last):
  File "test.py", line 185, in <module>
    outputs = loaded(input_ids.to("cpu"),attention_mask.to("cpu"))
  File "/home/ubuntu/anaconda3/envs/pytorch_latest_p36/lib/python3.6/site-packages/torch/nn/modules/module.py", line 550, in __call__
    result = self.forward(*input, **kwargs)
RuntimeError: The following operation failed in the TorchScript interpreter.
Traceback of TorchScript, serialized code (most recent call last):
  File "code/__torch__/transformers/modeling_bert.py", line 11, in forward
    attention_mask: Tensor) -> Tuple[Tensor]:
    _0 = self.dropout
    _1 = (self.bert).forward(input_ids, attention_mask, )
          ~~~~~~~~~~~~~~~~~~ <--- HERE
    _2 = (self.classifier).forward((_0).forward(_1, ), )
    return (_2,)
  File "code/__torch__/transformers/modeling_bert.py", line 34, in forward
    _10 = torch.to(extended_attention_mask, 6, False, False, None)
    attention_mask0 = torch.mul(torch.rsub(_10, 1., 1), CONSTANTS.c0)
    _11 = (_4).forward((_5).forward(input_ids, input, ), attention_mask0, )
                        ~~~~~~~~~~~ <--- HERE
    return (_3).forward(_11, )
class BertEmbeddings(Module):
  File "code/__torch__/transformers/modeling_bert.py", line 60, in forward
    _20 = (_16).forward(input_ids, )
    _21 = (_15).forward(input0, )
    _22 = (_14).forward(input, )
           ~~~~~~~~~~~~ <--- HERE
    input1 = torch.add(torch.add(_20, _21, alpha=1), _22, alpha=1)
    _23 = (_12).forward((_13).forward(input1, ), )
  File "code/__torch__/torch/nn/modules/sparse/___torch_mangle_1.py", line 7, in forward
  def forward(self: __torch__.torch.nn.modules.sparse.___torch_mangle_1.Embedding,
    input: Tensor) -> Tensor:
    token_type_embeddings = torch.embedding(self.weight, input, -1, False, False)
                            ~~~~~~~~~~~~~~~ <--- HERE
    return token_type_embeddings

Traceback of TorchScript, original code (most recent call last):
/home/ubuntu/anaconda3/envs/pytorch_latest_p36/lib/python3.6/site-packages/torch/nn/functional.py(1724): embedding
/home/ubuntu/anaconda3/envs/pytorch_latest_p36/lib/python3.6/site-packages/torch/nn/modules/sparse.py(114): forward
/home/ubuntu/anaconda3/envs/pytorch_latest_p36/lib/python3.6/site-packages/torch/nn/modules/module.py(534): _slow_forward
/home/ubuntu/anaconda3/envs/pytorch_latest_p36/lib/python3.6/site-packages/torch/nn/modules/module.py(548): __call__
/home/ubuntu/anaconda3/envs/pytorch_latest_p36/lib/python3.6/site-packages/transformers/modeling_bert.py(203): forward
/home/ubuntu/anaconda3/envs/pytorch_latest_p36/lib/python3.6/site-packages/torch/nn/modules/module.py(534): _slow_forward
/home/ubuntu/anaconda3/envs/pytorch_latest_p36/lib/python3.6/site-packages/torch/nn/modules/module.py(548): __call__
/home/ubuntu/anaconda3/envs/pytorch_latest_p36/lib/python3.6/site-packages/transformers/modeling_bert.py(792): forward
/home/ubuntu/anaconda3/envs/pytorch_latest_p36/lib/python3.6/site-packages/torch/nn/modules/module.py(534): _slow_forward
/home/ubuntu/anaconda3/envs/pytorch_latest_p36/lib/python3.6/site-packages/torch/nn/modules/module.py(548): __call__
/home/ubuntu/anaconda3/envs/pytorch_latest_p36/lib/python3.6/site-packages/transformers/modeling_bert.py(1279): forward
/home/ubuntu/anaconda3/envs/pytorch_latest_p36/lib/python3.6/site-packages/torch/nn/modules/module.py(534): _slow_forward
/home/ubuntu/anaconda3/envs/pytorch_latest_p36/lib/python3.6/site-packages/torch/nn/modules/module.py(548): __call__
/home/ubuntu/anaconda3/envs/pytorch_latest_p36/lib/python3.6/site-packages/torch/jit/__init__.py(1027): trace_module
/home/ubuntu/anaconda3/envs/pytorch_latest_p36/lib/python3.6/site-packages/torch/jit/__init__.py(875): trace
test.py(176): <module>
RuntimeError: Expected object of device type cuda but got device type cpu for argument #1 'self' in call to _th_index_select

**Comment by HamidShojanazeri** [comment]
@mfuntowicz I was wondering if there is any update/ workaround / PR  in progress to solve this issue. Thanks.

**Comment by HamidShojanazeri** [comment]
@chauhang Currently the HF model being traced on gpu can be run on machines with same one gpu. However the issue with tracing HF models on one gpu and running inference on another gpu in a multi-gpu setting is yet to be solved [issue#976](https://github.com/pytorch/serve/issues/976).

Logs of tracing Bert model on one gpu and running inference. 

[BERT_seq_cls_traced_one-gpu.logs.txt](https://github.com/pytorch/serve/files/6006652/BERT_seq_cls_traced_one-gpu.logs.txt)
